### ISSUE TITLE
Go binary size regression

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#26232]
Agent:
master: 110m
1.6: 74m
1.5: 68m
1.4: 38m

Pilot:
master: 108m 
1.6: 90m
1.5: 84m
1.4: 65m

Measured with 
```
for tag in master 1.6.0 1.5.0 1.4.0; do
    echo $tag
    git checkout $tag
    go build -o /tmp/agent ./pilot/cmd/pilot-discovery
    ll /tmp/agent
    echo
done
```

My theory is the 1.4 -> 1.5 regression in agent is for gateway SDS (pulls in k8s) and then 1.6 -> master regression is `google.golang.org/api/compute/v1`

cc @rshriram @kyessenov

**Comment by howardjohn** [comment]
istio-agent basically imports literally everything now since @costinm put an XDS server in it

**Comment by kyessenov** [comment]
I think that's strange that the client agent is larger than the service. Go doesn't do a good job at code elimination as well.

**Comment by kyessenov** [comment]
A minimalistic xDS server (e.g. go-control-plane test server) is around 22MB as a point of reference.

**Comment by howardjohn** [comment]
We also import literally every proto in xds (including v2 and v4) so we can handle envoyfilters which adds a lot

**Comment by ericvn** [comment]
@brian-avery Checking against 1.7 branch.

**Comment by mandarjog** [comment]
Here is the actual effect on memory.  Unshared ~= resident - shared.
TLDR;
1.7 Is using `~15 MB` per pod more for istio-agent, which I don't think is release blocking.

| Istio Version |  File size | Resident Memory | Unshared memory | Overage |
|--------|----|----|----|----|
| 1.5 | 68M | 55M | 15M | 0 |
| 1.6 | 74M |  55M | 17M | 2M |
| 1.7 | 110M | 80M | 30M | 15M |

Unshared memory costs increase **per pod**, whereas shared memory costs increase **per node**.

**Comment by howardjohn** [comment]
Can you share the commands you ran? Then we can compare some code changes
and see what makes improvements

**Comment by howardjohn** [comment]
btw -

go 1.14: 101M
go 1.15: 97M

so some small wins..

**Comment by howardjohn** [comment]
Notes to self: useful tools for analyzing:

```
goda graph -cluster -short ./pilot/cmd/pilot-agent | tee /tmp/goda
cat /tmp/goda | match 'tooltip.* color' | map 'i[9:-7]' | rg pilot/pkg/xds
```

```
goweight ./pilot/cmd/pilot-agent | tee /tmp/now | head -n 100
```

```
go build -o /tmp -ldflags '-extldflags -static -s -w' ./pilot/cmd/pilot-agent; ll /tmp/pilot-agent; file /tmp/pilot-agent; stat /tmp/pilot-agent
```

**Comment by Stono** [comment]
> 1.7 Is using ~15 MB per pod more for istio-agent, which I don't think is release blocking.

As someone using istio with a _lot_ of pods, a 25% increase in proxy usage, after we've just took a 10% increase 1.5 -> 1.6 for SDS, is certainly **upgrade** blocking for us.

**Comment by howardjohn** [comment]
Doing a bunch of work on https://github.com/howardjohn/istio/pull/new/agent/tiny-binary-full-branch to trim the deps a lot. Down to 37mb currently. 30mb seems feasible. It contains about 15 changes so I will be splitting it out into many smaller prs

**Comment by Stono** [comment]
That's awesome @howardjohn - guessing we're looking at 1.9 for most of them however?

**Comment by howardjohn** [comment]
Yeah unfortunately so. The bulk of the changes are from dropping Kubernetes imports which is not feasible on older versions.

**Comment by howardjohn** [comment]
Note sure how much this will impact once we are done trimming, but golang 1.16 will net us a few wins as well:
```
$ GOOS=linux GOARCH=amd64 LDFLAGS='-extldflags -static -s -w' common/scripts/gobuild.sh ./out/linux_amd64/ -tags="agent" ./pilot/cmd/pilot-agent; ll out/linux_amd64/pilot-agent

real    0m3.178s
user    0m5.407s
sys     0m0.816s
-rwxr-x--- 1 howardjohn primarygroup 61M Nov  9 09:33 out/linux_amd64/pilot-agent*
$ GOBINARY=/tmp/tmp.8k4Q6FfX0t/goroot/bin/go GOOS=linux GOARCH=amd64 LDFLAGS='-extldflags -static -s -w' common/scripts/gobuild.sh ./out/linux_amd64/ -tags="agent" ./pilot/cmd/pilot-agent; ll out/linux_amd64/pilot-agent

real    0m1.987s
user    0m3.605s
sys     0m0.735s
-rwxr-x--- 1 howardjohn primarygroup 55M Nov  9 09:33 out/linux_amd64/pilot-agent*
```

**Comment by howardjohn** [comment]
I am going with the assumption that we don't really care about istiod binary size and the agent is now a fraction of its size (38M), and considering this fixed
